go/libkb/pvl_interp.go

```diff
@@ -0,0 +1,825 @@
+					"DNS script cannot transform url")
+			case modeknown:
+				return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					"Script cannot transform after fetch")
+			}
+		default:
+			return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+				"Unsupported PVL instruction %d", i)
+		}
+	}
+
+	return nil
+}
+
+// Run each script on each TXT record of each domain.
+// Succeed if any succeed.
+func pvlRunDNS(g GlobalContextLite, scripts []*jsonw.Wrapper, startstate PvlScriptState) ProofError {
+	userdomain := startstate.Vars.Hostname
+	domains := []string{userdomain, "_keybase." + userdomain}
+	var firsterr ProofError
+	for _, d := range domains {
+		g.GetLog().Debug("Trying DNS: %v", d)
+
+		err := pvlRunDNSOne(g, scripts, startstate, d)
+		if err == nil {
+			return nil
+		}
+		if firsterr == nil {
+			firsterr = err
+		}
+	}
+
+	return firsterr
+}
+
+// Run each script on each TXT record of the domain.
+func pvlRunDNSOne(g GlobalContextLite, scripts []*jsonw.Wrapper, startstate PvlScriptState, domain string) ProofError {
+	txts, err := net.LookupTXT(domain)
+	if err != nil {
+		return NewProofError(keybase1.ProofStatus_DNS_ERROR,
+			"DNS failure for %s: %s", domain, err)
+	}
+
+	for _, record := range txts {
+		g.GetLog().Debug("For %s, got TXT record: %s", domain, record)
+
+		// Try all scripts.
+		for _, script := range scripts {
+			state := startstate
+			state.ActiveString = record
+			err = pvlRunScript(g, script, state)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	return NewProofError(keybase1.ProofStatus_NOT_FOUND,
+		"Checked %d TXT entries of %s, but didn't find signature",
+		len(txts), domain)
+}
+
+func pvlRunScript(g GlobalContextLite, script *jsonw.Wrapper, startstate PvlScriptState) ProofError {
+	var state = startstate
+	scriptlen, err := script.Len()
+	if err != nil {
+		return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Could not get length of script: %v", err)
+	}
+	if scriptlen < 1 {
+		return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+			"Empty scripts are not allowed.")
+	}
+	for i := 0; i < scriptlen; i++ {
+		ins := script.AtIndex(i)
+
+		// Sanity check.
+		if int(state.PC) != i {
+			return NewProofError(keybase1.ProofStatus_INVALID_PVL,
+				fmt.Sprintf("Execution failure, PC mismatch %v %v", state.PC, i))
+		}
+
+		newstate, perr := pvlStepInstruction(g, ins, state)
+		state = newstate
+		if perr != nil {
+			// Add additional info to INVALID_PVL errors.
+			if perr.GetProofStatus() == keybase1.ProofStatus_INVALID_PVL {
+				perr = NewProofError(keybase1.ProofStatus_INVALID_PVL,
+					fmt.Sprintf("Invalid PVL (%v): %v", state.PC, perr.GetDesc()))
+			}
+			return perr
+		}
+		state.PC++
+	}
+
+	// Script executed successfully with no errors.
+	return nil
+}
+
+func pvlStepInstruction(g GlobalContextLite, ins *jsonw.Wrapper, state PvlScriptState) (PvlScriptState, ProofError) {
+	for name, step := range PvlSteps {
+		if pvlJSONHasKey(ins, name) {
```

Comment:
Is this the place where we need to prohibit multiple instruction names in a single `ins` map?